onionbalance/hs_v3/service.py

```diff
@@ -294,67 +316,102 @@ def _publish_descriptor(self, is_first_desc):
+            distinct_descriptors = False
 
-        try:
-            desc = descriptor.OBDescriptor(self.onion_address, self.identity_priv_key,
-                                           blinding_param, intro_points, is_first_desc)
-        except descriptor.BadDescriptor:
-            return
-
-        logger.info("Service %s created %s descriptor (%s intro points) (blinding param: %s) (size: %s bytes). About to publish:",
-                    self.onion_address, "first" if is_first_desc else "second",
-                    len(desc.intro_set), blinding_param.hex(), len(str(desc.v3_desc)))
-
-        # When we do a v3 HSPOST on the control port, Tor decodes the
-        # descriptor and extracts the blinded pubkey to be used when uploading
-        # the descriptor. So let's do the same to compute the responsible
-        # HSDirs:
-        blinded_key = desc.get_blinded_key()
-
-        # Calculate responsible HSDirs for our service
-        try:
-            responsible_hsdirs = hashring.get_responsible_hsdirs(blinded_key, is_first_desc)
-        except hashring.EmptyHashRing:
-            logger.warning("Can't publish desc with no hash ring. Delaying...")
-            return
 
-        desc.set_last_publish_attempt_ts(datetime.datetime.utcnow())
-
-        logger.info("Uploading %s descriptor for %s to %s",
-                    "first" if is_first_desc else "second",
-                    self.onion_address, responsible_hsdirs)
+        if distinct_descriptors:
+        # Derive blinding parameter
+            _, time_period_number = hashring.get_srv_and_time_period(is_first_desc)
+            blinding_param = my_onionbalance.consensus.get_blinding_param(self._get_identity_pubkey_bytes(), time_period_number)
+            
+            
+            try:
+                desc = descriptor.OBDescriptor(self.onion_address, self.identity_priv_key, blinding_param, [], is_first_desc)
+            except descriptor.BadDescriptor:
+                return
+            blinded_key = desc.get_blinded_key()
+            try:
+                responsible_hsdirs = hashring.get_responsible_hsdirs(blinded_key, is_first_desc)
+            except hashring.EmptyHashRing:
+                logger.warning("Can't publish desc with no hash ring. Delaying...")
+                return
+            for hsdir in responsible_hsdirs:   
+                try:
+                    intro_points = self._get_intros_for_distinct_desc(max_intro_points)
+                except NotEnoughIntros:
+                    return
+                logger.info("uploading a distinct descriptor")
+                try:
+                    desc = descriptor.OBDescriptor(self.onion_address, self.identity_priv_key,
+                                                   blinding_param, intro_points, is_first_desc)
+                except descriptor.BadDescriptor:
+                    return
+                self._upload_descriptor(my_onionbalance.controller.controller, desc, hsdir)
+        else:
```

Comment:
So this is the non-distinct descriptor mode? is this identical to the previous functionality of Onionbalance?
In any case, let's move this also to its own function and document how it works.